core/src/main/java/org/jdbi/v3/core/async/JdbiExecutorImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.async;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+import org.jdbi.v3.core.HandleCallback;
+import org.jdbi.v3.core.HandleConsumer;
+import org.jdbi.v3.core.Jdbi;
+import org.jdbi.v3.core.extension.ExtensionCallback;
+import org.jdbi.v3.core.extension.ExtensionConsumer;
+import org.jdbi.v3.core.transaction.TransactionIsolationLevel;
+
+public class JdbiExecutorImpl implements JdbiExecutor {
+
+    private final Jdbi jdbi;
+    private final Executor executor;
+
+    /**
+     * Construct a {@link JdbiExecutor}.
+     *
+     * <p>
+     * The executor to pass in needs to be sized to the use case. A rule of thumb is to
+     * have the max number of workers be equal to the max number of connections in the
+     * connection pool. The worker queue in the executor should probably be bounded, unless
+     * the caller(s) already has a bound for the number of outstanding requests. Making the
+     * queue bounded will mean you are blocking the calling thread when the queue fills up.
+     * Whether or not that is a
+     * </p>
+     *
+     * @param jdbi the initialized Jdbi class
+     * @param executor an executor to use for all database calls
+     */
+    public JdbiExecutorImpl(Jdbi jdbi, Executor executor) {
```

Comment:
Rather than have a `Impl` class be public, it would be nice to have the `Impl` class be package-private and provide a static factory method like `JdbiExecutor.create(...)`

Reply:
The nice thing about the Impl is that it can be overridden by only implementing the one method and chain the JdbiExecutor (so you can for example have a JdbiExecutorMetrics). One alternative is to make all the methods in the JdbiExecutor interface defaults, and have the defaults implementation all redirect to the one doExecute method. Same end result, but then you can make the Impl package private without losing the capability to make an easy delegate. At the cost of exposing a bit more details in the interface.

Would look like:
```java
public interface JdbiExecutor {
    /**
     * Single delegate handler to which all other methods delegate
     * ...
     */
    <R, X extends Exception> CompletionStage<R> doExecute(final Handler<R, X> handler);

    /**
     * A convenience function which manages the lifecycle of a handle and yields it to a callback
     * for use by clients.
     * ...
     */
    default <R, X extends Exception> CompletionStage<R> withHandle(HandleCallback<R, X> callback) {
        return doExecute(jdbi -> jdbi.withHandle(callback));
    }
```

Reply:
That is a fair point. Unless Henning has a strong opinion, I guess I will leave it up to you which style to prefer.

Reply:
I've changed it to use a public abstract class AbstractJdbiExecutor that does the redirect to the single method and then a JdbiExecutor.create which instantiates a package private JdbiExecutorImpl which only implements that single method.

This way, it maintains the clean interface, but still allows easy chaining of the implementation.